src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf16Utility.Validation.cs

```diff
@@ -112,16 +129,33 @@ static Utf16Utility()
 
 #if DEBUG
                         // Quick check to ensure we didn't accidentally set the 0x8000 bit of any element.
-                        uint debugMask = (uint)Sse2.MoveMask(charIsNonAscii.AsByte());
+                        uint debugMask;
+                        if (AdvSimd.Arm64.IsSupported)
+                        {
+                            debugMask = (uint)Arm64MoveMask(charIsNonAscii.AsByte());
+                        }
+                        else
+                        {
+                            debugMask = (uint)Sse2.MoveMask(charIsNonAscii.AsByte());
+                        }
                         Debug.Assert((debugMask & 0b_1010_1010_1010_1010) == 0, "Shouldn't have set the 0x8000 bit of any element in 'charIsNonAscii'.");
 #endif // DEBUG
 
                         // Sets the 0x8080 bits of each element in 'charIsNonAscii' if the corresponding
                         // input was 0x0800 <= [value]. This also handles the missing range a few lines above.
 
-                        Vector128<ushort> charIsThreeByteUtf8Encoded = Sse2.Subtract(vectorZero, Sse2.ShiftRightLogical(utf16Data, 11));
+                        Vector128<ushort> charIsThreeByteUtf8Encoded;
+                        if (AdvSimd.Arm64.IsSupported)
+                        {
+                            charIsThreeByteUtf8Encoded = AdvSimd.Arm64.Subtract(vectorZero.AsDouble(), AdvSimd.ShiftRightLogical(utf16Data, 11).AsDouble()).AsUInt16();
```

Comment:
This `Subtract` method is receiving doubles, so I'm unsure if it will do the same as the `Sse2` version. There are other `Subtract*` methods, although they receive `Vector64`. I suspect those are the appropriate options to use?

Reply:
There is variant of `Substract` that takes `Vector128<short>` - [https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs#L13965](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs#L13965)

---
In reply to: [452955480](https://github.com/dotnet/runtime/pull/39050#discussion_r452955480) [](ancestors = 452955480)

Reply:
Oh you're right, I was only looking in Arm64. Thanks.